# Ex1 - Filtering and Sorting Data



### Step 1. Import the necessary libraries

In [1]:
import pandas as pd
import numpy as np

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv) and assign it to a variable called chipo.

In [2]:
url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv"
chipo = pd.read_csv(url, sep="\t")
chipo

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

### Step 3. Name of the max valued product

In [3]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


In [26]:
chipo.describe()

order_id     quantity   item_price
count  4622.000000  4622.000000  4622.000000
mean    927.254868     1.075725     7.464336
std     528.890796     0.410186     4.245557
min       1.000000     1.000000     1.090000
25%     477.250000     1.000000     3.390000
50%     926.000000     1.000000     8.750000
75%    1393.000000     1.000000     9.250000
max    1834.000000    15.000000    44.250000

In [4]:
chipo["item_price"]= chipo["item_price"].str.replace("$","").astype(float)
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.7+ KB


In [5]:
chipo.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

In [6]:
chipo["item_price"].max()
maximo = chipo["item_price"].max()
maximo

np.float64(44.25)

In [7]:
str(chipo[chipo["item_price"] ==chipo["item_price"].max()]["item_name"].values[0])

'Chips and Fresh Tomato Salsa'

In [8]:
chipo["item_name"].value_counts()

item_name
Chicken Bowl                             726
Chicken Burrito                          553
Chips and Guacamole                      479
Steak Burrito                            368
Canned Soft Drink                        301
Chips                                    211
Steak Bowl                               211
Bottled Water                            162
Chicken Soft Tacos                       115
Chips and Fresh Tomato Salsa             110
Chicken Salad Bowl                       110
Canned Soda                              104
Side of Chips                            101
Veggie Burrito                            95
Barbacoa Burrito                          91
Veggie Bowl                               85
Carnitas Bowl                             68
Barbacoa Bowl                             66
Carnitas Burrito                          59
Steak Soft Tacos                          55
6 Pack Soft Drink                         54
Chips and Tomatillo Red Chili Salsa       48


In [9]:
chipo[chipo["item_name"] == "Chicken Bowl"]

order_id  quantity     item_name  \
4            2         2  Chicken Bowl   
5            3         1  Chicken Bowl   
13           7         1  Chicken Bowl   
19          10         1  Chicken Bowl   
26          13         1  Chicken Bowl   
...        ...       ...           ...   
4590      1825         1  Chicken Bowl   
4591      1825         1  Chicken Bowl   
4595      1826         1  Chicken Bowl   
4599      1827         1  Chicken Bowl   
4604      1828         1  Chicken Bowl   

                                     choice_description  item_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
13    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
19    [Tomatillo Red Chili Salsa, [Fajita Vegetables...        8.75  
26    [Roasted Chili Corn Salsa (Medium), [Pinto Bea...        8.49  
...                                                 ...         ...  
4590  [Roasted Chili Corn Salsa, [Rice, Black Beans,...       11.25  
4591  [Tomatillo Red Chili Salsa, [Rice, Black Beans...        8.75  
4595  [Tomatillo Green Chili Salsa, [Rice, Black Bea...        8.75  
4599      [Roasted Chili Corn Salsa, [Cheese, Lettuce]]        8.75  
4604  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...        8.75  

[726 rows x 5 columns]

In [10]:
len(chipo[chipo["item_name"] == "Chicken Bowl"])

726

In [11]:
chipo["item_name"].unique()

array(['Chips and Fresh Tomato Salsa', 'Izze', 'Nantucket Nectar',
       'Chips and Tomatillo-Green Chili Salsa', 'Chicken Bowl',
       'Side of Chips', 'Steak Burrito', 'Steak Soft Tacos',
       'Chips and Guacamole', 'Chicken Crispy Tacos',
       'Chicken Soft Tacos', 'Chicken Burrito', 'Canned Soda',
       'Barbacoa Burrito', 'Carnitas Burrito', 'Carnitas Bowl',
       'Bottled Water', 'Chips and Tomatillo Green Chili Salsa',
       'Barbacoa Bowl', 'Chips', 'Chicken Salad Bowl', 'Steak Bowl',
       'Barbacoa Soft Tacos', 'Veggie Burrito', 'Veggie Bowl',
       'Steak Crispy Tacos', 'Chips and Tomatillo Red Chili Salsa',
       'Barbacoa Crispy Tacos', 'Veggie Salad Bowl',
       'Chips and Roasted Chili-Corn Salsa',
       'Chips and Roasted Chili Corn Salsa', 'Carnitas Soft Tacos',
       'Chicken Salad', 'Canned Soft Drink', 'Steak Salad Bowl',
       '6 Pack Soft Drink', 'Chips and Tomatillo-Red Chili Salsa', 'Bowl',
       'Burrito', 'Crispy Tacos', 'Carnitas Crispy Tacos

In [12]:
chipo[chipo["item_name"]=="Chicken Burrito"]

order_id  quantity        item_name  \
16           8         1  Chicken Burrito   
17           9         1  Chicken Burrito   
23          12         1  Chicken Burrito   
29          15         1  Chicken Burrito   
45          21         1  Chicken Burrito   
...        ...       ...              ...   
4556      1811         1  Chicken Burrito   
4558      1812         1  Chicken Burrito   
4559      1812         1  Chicken Burrito   
4566      1816         1  Chicken Burrito   
4569      1817         1  Chicken Burrito   

                                     choice_description  item_price  
16    [Tomatillo-Green Chili Salsa (Medium), [Pinto ...        8.49  
17    [Fresh Tomato Salsa (Mild), [Black Beans, Rice...        8.49  
23    [[Tomatillo-Green Chili Salsa (Medium), Tomati...       10.98  
29    [Tomatillo-Green Chili Salsa (Medium), [Pinto ...        8.49  
45    [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       10.98  
...                                                 ...         ...  
4556  [Tomatillo Green Chili Salsa, [Fajita Vegetabl...       11.25  
4558  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...        8.75  
4559  [Tomatillo Red Chili Salsa, [Rice, Cheese, Gua...       11.25  
4566  [Roasted Chili Corn Salsa, [Pinto Beans, Chees...        8.75  
4569    [Fresh Tomato Salsa, [Fajita Vegetables, Rice]]        8.75  

[553 rows x 5 columns]

### Step 4. How many products cost more than $10.00?

In [13]:
chipo[chipo["item_price"]>10]

order_id  quantity           item_name  \
4            2         2        Chicken Bowl   
5            3         1        Chicken Bowl   
7            4         1       Steak Burrito   
13           7         1        Chicken Bowl   
23          12         1     Chicken Burrito   
...        ...       ...                 ...   
4610      1830         1       Steak Burrito   
4611      1830         1      Veggie Burrito   
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   

                                     choice_description  item_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
13    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
23    [[Tomatillo-Green Chili Salsa (Medium), Tomati...       10.98  
...                                                 ...         ...  
4610  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4611  [Tomatillo Green Chili Salsa, [Rice, Fajita Ve...       11.25  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...       11.25  

[1130 rows x 5 columns]

In [14]:
chipo[chipo["item_price"]>10]["item_price"]

4       16.98
5       10.98
7       11.75
13      11.25
23      10.98
        ...  
4610    11.75
4611    11.25
4617    11.75
4618    11.75
4619    11.25
Name: item_price, Length: 1130, dtype: float64

In [16]:
chipo[chipo["item_price"]>10]["item_price"].sum()

np.float64(14003.54)

### Step 4.1: Y cuántos pedidos se han hecho con un producto de más de 10$? Es lo mismo?

In [112]:
chipo[chipo["item_price"]>10]["item_price"].count()

np.int64(1130)

### Step 4.2: Y cuántos pedidos se han hecho de más de 10$? Es lo mismo?

In [27]:
chipo.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

El resultado debería ser el total de pedidos, ya que mirando el describe del conjunto obtenemos la siguiente información:


In [ ]:
chipo[chipo["item_price"]>10]["item_price"].describe()

count    1130.000000
mean       12.392513
std         3.199968
min        10.500000
25%        11.250000
50%        11.250000
75%        11.750000
max        44.250000
Name: item_price, dtype: float64

Como observamos aquí, la compra total mínima realizada es de 10.5

In [35]:
gasto_pedido = chipo.groupby("order_id",as_index="Gasto total").item_price.sum()
print(gasto_pedido)

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_price, Length: 1834, dtype: float64


In [38]:
print(gasto_pedido[gasto_pedido > 10].count())

1834


### Step 4.3: Y en cuántos pedidos se ha pagado más de 10$ por un mismo producto? Es lo mismo?

In [49]:
len(chipo.order_id[chipo.item_price > 10].unique())

863

### Step 5. What is the price of each item and name it unit_price. Get only item_name and unit_price

In [ ]:
chipo.groupby("item_name")[["item_price"]].mean()
"""
aux = pd.DataFrame(chipo.groupby("item_name")[["item_price"]].mean())
aux.item_price.rename("unit_price")
aux
"""

item_price
item_name                                        
6 Pack Soft Drink                        6.610185
Barbacoa Bowl                           10.187273
Barbacoa Burrito                         9.832418
Barbacoa Crispy Tacos                   10.928182
Barbacoa Salad Bowl                     10.640000
Barbacoa Soft Tacos                     10.018400
Bottled Water                            1.867654
Bowl                                    14.800000
Burrito                                  7.400000
Canned Soda                              1.320577
Canned Soft Drink                        1.457641
Carnitas Bowl                           10.833971
Carnitas Burrito                        10.132712
Carnitas Crispy Tacos                   11.137143
Carnitas Salad                           8.990000
Carnitas Salad Bowl                     11.056667
Carnitas Soft Tacos                      9.398500
Chicken Bowl                            10.113953
Chicken Burrito                         10.082857
Chicken Crispy Tacos                    10.045319
Chicken Salad                            9.010000
Chicken Salad Bowl                      11.170455
Chicken Soft Tacos                       9.635565
Chips                                    2.342844
Chips and Fresh Tomato Salsa             3.285091
Chips and Guacamole                      4.595073
Chips and Mild Fresh Tomato Salsa        3.000000
Chips and Roasted Chili Corn Salsa       3.084091
Chips and Roasted Chili-Corn Salsa       2.390000
Chips and Tomatillo Green Chili Salsa    3.087209
Chips and Tomatillo Red Chili Salsa      3.072917
Chips and Tomatillo-Green Chili Salsa    2.544194
Chips and Tomatillo-Red Chili Salsa      2.987500
Crispy Tacos                             7.400000
Izze                                     3.390000
Nantucket Nectar                         3.641111
Salad                                    7.400000
Side of Chips                            1.840594
Steak Bowl                              10.711801
Steak Burrito                           10.465842
Steak Crispy Tacos                      10.209714
Steak Salad                              8.915000
Steak Salad Bowl                        11.847931
Steak Soft Tacos                         9.746364
Veggie Bowl                             10.211647
Veggie Burrito                           9.839684
Veggie Crispy Tacos                      8.490000
Veggie Salad                             8.490000
Veggie Salad Bowl                       10.138889
Veggie Soft Tacos                       10.565714

### Step 6. Sort by the name of the item

In [79]:
aux = chipo.groupby("item_name")[["item_price"]].mean().sort_values("item_name")
aux

item_price
item_name                                        
6 Pack Soft Drink                        6.610185
Barbacoa Bowl                           10.187273
Barbacoa Burrito                         9.832418
Barbacoa Crispy Tacos                   10.928182
Barbacoa Salad Bowl                     10.640000
Barbacoa Soft Tacos                     10.018400
Bottled Water                            1.867654
Bowl                                    14.800000
Burrito                                  7.400000
Canned Soda                              1.320577
Canned Soft Drink                        1.457641
Carnitas Bowl                           10.833971
Carnitas Burrito                        10.132712
Carnitas Crispy Tacos                   11.137143
Carnitas Salad                           8.990000
Carnitas Salad Bowl                     11.056667
Carnitas Soft Tacos                      9.398500
Chicken Bowl                            10.113953
Chicken Burrito                         10.082857
Chicken Crispy Tacos                    10.045319
Chicken Salad                            9.010000
Chicken Salad Bowl                      11.170455
Chicken Soft Tacos                       9.635565
Chips                                    2.342844
Chips and Fresh Tomato Salsa             3.285091
Chips and Guacamole                      4.595073
Chips and Mild Fresh Tomato Salsa        3.000000
Chips and Roasted Chili Corn Salsa       3.084091
Chips and Roasted Chili-Corn Salsa       2.390000
Chips and Tomatillo Green Chili Salsa    3.087209
Chips and Tomatillo Red Chili Salsa      3.072917
Chips and Tomatillo-Green Chili Salsa    2.544194
Chips and Tomatillo-Red Chili Salsa      2.987500
Crispy Tacos                             7.400000
Izze                                     3.390000
Nantucket Nectar                         3.641111
Salad                                    7.400000
Side of Chips                            1.840594
Steak Bowl                              10.711801
Steak Burrito                           10.465842
Steak Crispy Tacos                      10.209714
Steak Salad                              8.915000
Steak Salad Bowl                        11.847931
Steak Soft Tacos                         9.746364
Veggie Bowl                             10.211647
Veggie Burrito                           9.839684
Veggie Crispy Tacos                      8.490000
Veggie Salad                             8.490000
Veggie Salad Bowl                       10.138889
Veggie Soft Tacos                       10.565714

### Step 7. What was the quantity of the most expensive item ordered? 2 ways

In [103]:
aux["quantity"] = chipo.groupby("item_name")[["quantity"]].sum()
aux

item_price  quantity
item_name                                                  
6 Pack Soft Drink                        6.610185        55
Barbacoa Bowl                           10.187273        66
Barbacoa Burrito                         9.832418        91
Barbacoa Crispy Tacos                   10.928182        12
Barbacoa Salad Bowl                     10.640000        10
Barbacoa Soft Tacos                     10.018400        25
Bottled Water                            1.867654       211
Bowl                                    14.800000         4
Burrito                                  7.400000         6
Canned Soda                              1.320577       126
Canned Soft Drink                        1.457641       351
Carnitas Bowl                           10.833971        71
Carnitas Burrito                        10.132712        60
Carnitas Crispy Tacos                   11.137143         8
Carnitas Salad                           8.990000         1
Carnitas Salad Bowl                     11.056667         6
Carnitas Soft Tacos                      9.398500        40
Chicken Bowl                            10.113953       761
Chicken Burrito                         10.082857       591
Chicken Crispy Tacos                    10.045319        50
Chicken Salad                            9.010000         9
Chicken Salad Bowl                      11.170455       123
Chicken Soft Tacos                       9.635565       120
Chips                                    2.342844       230
Chips and Fresh Tomato Salsa             3.285091       130
Chips and Guacamole                      4.595073       506
Chips and Mild Fresh Tomato Salsa        3.000000         1
Chips and Roasted Chili Corn Salsa       3.084091        23
Chips and Roasted Chili-Corn Salsa       2.390000        18
Chips and Tomatillo Green Chili Salsa    3.087209        45
Chips and Tomatillo Red Chili Salsa      3.072917        50
Chips and Tomatillo-Green Chili Salsa    2.544194        33
Chips and Tomatillo-Red Chili Salsa      2.987500        25
Crispy Tacos                             7.400000         2
Izze                                     3.390000        20
Nantucket Nectar                         3.641111        29
Salad                                    7.400000         2
Side of Chips                            1.840594       110
Steak Bowl                              10.711801       221
Steak Burrito                           10.465842       386
Steak Crispy Tacos                      10.209714        36
Steak Salad                              8.915000         4
Steak Salad Bowl                        11.847931        31
Steak Soft Tacos                         9.746364        56
Veggie Bowl                             10.211647        87
Veggie Burrito                           9.839684        97
Veggie Crispy Tacos                      8.490000         1
Veggie Salad                             8.490000         6
Veggie Salad Bowl                       10.138889        18
Veggie Soft Tacos                       10.565714         8

In [104]:
aux.quantity[aux.item_price == aux.item_price.max()]

item_name
Bowl    4
Name: quantity, dtype: int64

In [105]:
aux.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50 entries, 6 Pack Soft Drink to Veggie Soft Tacos
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   item_price  50 non-null     float64
 1   quantity    50 non-null     int64  
dtypes: float64(1), int64(1)
memory usage: 3.2+ KB


### Step 8. How many times was a Veggie Salad Bowl ordered?

In [109]:

aux["quantity"][aux.index == "Veggie Salad Bowl"]

item_name
Veggie Salad Bowl    18
Name: quantity, dtype: int64

### Step 9. How many times did someone order more than one Canned Soda?

In [97]:
chipo.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  unit_price  
0                                                NaN        2.39         NaN  
1                                       [Clementine]        3.39         NaN  
2                                            [Apple]        3.39         NaN  
3                                                NaN        2.39         NaN  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98         NaN

In [98]:
chipo.order_id[(chipo.quantity >= 2) & (chipo.item_name == "Canned Soda")].count()

np.int64(20)